cmd/protoc-gen-go-grpc/README.md

```diff
@@ -4,18 +4,115 @@ This tool generates Go language bindings of `service`s in protobuf definition
 files for gRPC.  For usage information, please see our [quick start
 guide](https://grpc.io/docs/languages/go/quickstart/).
 
-## Future-proofing services
+## Service implementation and registration
 
-By default, to register services using the methods generated by this tool, the
-service implementations must embed the corresponding
-`Unimplemented<ServiceName>Server` for future compatibility.  This is a behavior
-change from the grpc code generator previously included with `protoc-gen-go`.
-To restore this behavior, set the option `requireUnimplementedServers=false`.
-E.g.:
+**NOTE:** service registration has changed from the previous version of the
+code generator.  Please read this section carefully if you are migrating.
 
+To register your service handlers with a gRPC server, first implement the
+methods as either functions or methods on a struct.  Examples:
+
+```go
+// As a function:
+
+func unaryEcho(ctx context.Context, req *pb.EchoRequest) (*pb.EchoResponse, error) {
+     // Echo.UnaryEcho implementation
+}
+
+// As a struct + method:
+
+type myEchoService struct {
+     // ...fields used by this service...
+}
+
+func (s *myEchoService) UnaryEcho(ctx context.Context, req *pb.EchoRequest) (*pb.EchoResponse, error) {
+     // Echo.UnaryEcho implementation
+}
+```
+
+Then create an instance of the generated `Service` struct type and initialize
+the handlers which have been implemented:
+
+```go
+func main() {
+     // ...
+
+     // As a function:
+     echoService := pb.EchoService{
+          UnaryEcho: unaryEcho,
+          // etc
+     }
+
+     // As a struct+method:
+     mes := &myEchoService{...}
+     echoService := pb.EchoService{
+          UnaryEcho: mes.UnaryEcho,
+          // etc
+     }
+
+     // ...
+ }
 ```
-  protoc --go-grpc_out=requireUnimplementedServers=false[,other options...]:. \
+
+Finally, pass this `Service` instance to the generated `Register` function:
+
+```go
+     pb.RegisterEchoService(grpcServer, echoService)
 ```
 
-Note that this is not recommended, and the option is only provided to restore
-backward compatibility with previously-generated code.
+### Migration from older version
+
+The previous version of protoc-gen-go-grpc used a different method to register
+services.  In that version, it was only possible to register a service
+implementation that was a complete implementation of the service.  It was also
+possible to embed an `Unimplemented` implementation of the service, which was
+also generated and returned an UNIMPLEMENTED status for all methods.  To make
+it easier to migrate from the previous version, two new symbols are generated:
+`New<Service>Service` and `Unstable<Service>Service`.
+
+`New<Service>Service` allows for easy wrapping of a service implementation into
+an instance of the new generated `Service` struct type.  *This has drawbacks,
+however: `New<Service>Service` accepts any parameter, and does not panic or
```

Comment:
If methods are missing or misspelt, does it eventually return UNIMPLEMENTED at runtime? Maybe document what the expected behavior is in that case.

Reply:
I just the generated code, which does say that missing/misspelt methods will return UNIMPLEMENTED. Maybe it makes sense to document that here as well.

Reply:
Done